[ONNX] Introduce Input/Ouptut adapter; Switch to 'DynamoExporter'

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99202
* #99191
* __->__ #98421

Summary
* Introduce input/output adapter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output adapter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

BowenBao committed:
[ONNX] Introduce Input/Ouptut formatter; Switch to 'DynamoExporter'

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -163,21 +165,94 @@ def serialize(
 
+    @property
+    def input_formatter(self) -> InputFormatter:
+        """The input formatter used to convert inputs between format of PyTorch, torch.fx and ONNX."""
```
todo: update a bit outdated comment.

BowenBao commented on torch/onnx/_internal/fx/dynamo_exporter.py:
```diff
@@ -28,41 +37,156 @@ def export(self) -> torch.onnx.ExportOutput:
         compiler = GraphCaptureCompiler()
         torch._dynamo.reset()
-        torch._dynamo.optimize(compiler.compile, nopython=True)(self.model)(*bound_args)
+        # TODO(titaiwang): Set `dynamic` according to `self.options.dynamic_shapes`
```
Unrelated to this PR. I added this todo and I'm curious if it is accurate @titaiwangms ? It felt strange that export option `dynamic_shapes` does not affect dynamo `dynamic`/`tracing_mode`.

titaiwangms replied:
I think this kind of circle back to the question that "Do we have to align dynamo and make_fx mode"? Based on my findings, we can pass dynamic test cases without setting this one dynamic, but I can try enable this.

BowenBao replied:
I think it certainly affects dynamo's behavior in capturing fx graphs with dynamic shapes. I suspect without alignment, we are only doing "dynamo static capture" + "refill symbolic shapes".

I feel turning this on in dynamo call won't be a trivial task. I can imagine it introducing a lot more graph breaks, which is resolvable with the new `constraints` api FYI https://github.com/pytorch/pytorch/pull/96699 if you are interested, and perhaps other patterns. Let's dig more and do this when we reach the step of refining `dynamic` apis for `dynamo_export`.

titaiwangms replied:
Thanks for the info!

titaiwangms replied:
I will be looking into this now.

titaiwangms replied:
https://github.com/pytorch/pytorch/issues/98962

BowenBao commented on torch/onnx/_internal/fx/dynamo_exporter.py:
```diff
@@ -28,41 +37,156 @@ def export(self) -> torch.onnx.ExportOutput:
         return self.export_fx_to_onnx(
-            compiler.captured_graph, tuple(arg for arg in bound_args if arg is not None)
+            compiler.captured_graph,
+            model_args,
```
```suggestion
            model_args
```
nit

BowenBao commented on torch/onnx/_internal/fx/dynamo_exporter.py:
```diff
@@ -28,41 +37,156 @@ def export(self) -> torch.onnx.ExportOutput:
-class DynamoExportExporter(torch.onnx._internal.fx.fx_exporter.FXGraphModuleExporter):
+    def __init__(
+        self,
+        model: torch.nn.Module,
```
```suggestion
        model: Union[torch.nn.Module, Callable],
```

BowenBao committed:
documentations and small tweaks on "[ONNX] Introduce Input/Ouptut formatter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output formatter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output formatter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -163,21 +164,94 @@ def serialize(
         return self._model_proto
 
+    @property
+    def input_formatter(self) -> InputFormatter:
```
I thought about exposing method api instead of the object to do formatting. But then I think maybe object is better for future extension such as reverse formatting etc.

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -221,14 +289,13 @@ def __init__(
             self.options = options
         assert self.options is not None
 
-        self.model: torch.nn.Module = (
```
I'm not sure if this is needed. For this I could revert.

At least the pytree flatten part inside `WrappedFuncModule` is not needed anymore.

BowenBao committed:
a few more comments and proofreading on "[ONNX] Introduce Input/Ouptut formatter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output formatter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output formatter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

BowenBao committed:
fix public interface CI; More docstrings on "[ONNX] Introduce Input/Ouptut formatter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output formatter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output formatter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

BowenBao committed:
remove debug code change on "[ONNX] Introduce Input/Ouptut formatter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output formatter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output formatter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

BowenBao committed:
update doc on "[ONNX] Introduce Input/Ouptut formatter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output formatter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output formatter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Introduce Input/Ouptut formatter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output formatter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output formatter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Introduce Input/Ouptut formatter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output formatter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output formatter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/exporter.py:
```diff
@@ -163,21 +164,194 @@ def serialize(
 
 
+# TODO(bowbao): Add diagnostics for IO formatters.
+class InputFormatStep(abc.ABC):
```
Have you considered making it a protocol?

BowenBao replied:
That's a good idea.

BowenBao committed:
Resolve huge conflict after rebase on "[ONNX] Introduce Input/Ouptut formatter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output formatter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output formatter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

thiagocrepaldi reviewed (commented):
Having input and output formatter is a good way to help users to bridge the difference between the pytorch and onnx models!

Maybe we could find a way to hide all this complexity and only expose a helper class which does the dirty trick. Below is an idea of what it could looks like


```python
from torch.onnx._internals import input_formatter, output_formatter


class ONNX2PytorchOutput(torch.nn.Module):
    def __init__(self, torch_model: torch.nn.Module, ort_inference_session: onnxruntime.InferenceSession):
        """Facilitates ONNX model execution using PyTorch's input/output signature.

        After the model is exported to ONNX, its inputs and outputs may be different
        than the original PyTorch model's input/output.

        As example, Pytorch's `None` inputs are ignored by ONNX models.
        Another example is that keyword arguments are converted into positional arguments by ONNX models.
        """
        super().__init__()
        self.torch_model = torch_model
        self.ort_inference_session = ort_inference_session

    def forward(self, *torch_args, **torch_kwargs):
        """Executes ONNX model using PyTorch's input/output signature.

        Internally, Pytorch arguments are automatically parsed into the expected ONNX format,
        next the ONNX model is executed using ONNXRuntime, and its output
        is parsed back into the original PyTorch output format."""
        onnx_input = input_formatter.to_onnx(*torch_args, **torch_kwargs)
        onnx_output = self.ort_inference_session.run(None, onnx_input)
        torch_output = output_formatter.to_onnx(onnx_output)
        return torch_output
```

Another improvement we could make is to actually use ONNX's [ModelProto.doc_string](https://github.com/onnx/onnx/blob/51eac7e73fc97ff1a78396a2e94c7b1775859cef/onnx/onnx.proto3#L381) to *also* document the input/output format this model expects. [ValueInfoProto.doc_string](https://github.com/onnx/onnx/blob/main/onnx/onnx.proto3#L183) could also play a role in this. In this way, the ONNX model is self-contained in the sense it provides users insights on how the model expects to be invoked and what it produces

BowenBao committed:
adjust public api on "[ONNX] Introduce Input/Ouptut formatter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output formatter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output formatter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Introduce Input/Ouptut formatter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output formatter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output formatter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/fx/fx_symbolic_exporter.py:
```diff
@@ -208,12 +188,7 @@ def export(self) -> FXSymbolicTraceExportOutput:
 
         export_output = self.export_fx_to_onnx(
-            graph_module, bound_args + replaced_attrs
+            graph_module, tuple(bound_args) + replaced_attrs
```
```suggestion
            graph_module, (*bound_args, *replaced_attrs)
```

BowenBao committed:
fix CI docs and onnx test on "[ONNX] Introduce Input/Ouptut formatter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output formatter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output formatter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/fx_exporter.py:
```diff
@@ -39,6 +311,17 @@ def export_fx_to_onnx(
             )
+            # ONNX does not support None inputs. During graph building, all None inputs
+            # are removed. Here we register this step to input formatter.
+            self._apply_input_format_step(
```
Ideally it's bottom up: If a pass modifies the input/output format, it updates the step into the formatter. It is not there yet in this PR.

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -334,9 +517,9 @@ def dynamo_export(
     _assert_dependencies(resolved_export_options)
 
-    from torch.onnx._internal.fx.dynamo_exporter import DynamoOptimizeExporter
+    from torch.onnx._internal.fx.dynamo_exporter import DynamoExporter
```
We don't want to introduce onnx/onnx-script dependency to torch.

justinchuby replied:
Was the dependency from dynamo_exporter? May be worth commenting here to remind readers

BowenBao replied:
done

BowenBao committed:
Update on "[ONNX] Introduce Input/Ouptut formatter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output formatter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output formatter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/fx/dynamo_exporter.py:
```diff
@@ -1,25 +1,37 @@
 
 
-class DynamoOptimizeExporter(torch.onnx._internal.fx.fx_exporter.FXGraphModuleExporter):
+class DynamoOptimizeExporter(fx_exporter.FXGraphModuleExporter):
```
My previous comments were based on the fact that the I/O adapters were meant as a helper for users, but now I see that you are also using it internally, right before the X graph extractors. Each FX extractor with a different set of "steps".

For the public use, I am inclined in having it as a separate class helper, away from `ExportOutput` or any other `torch.onnx` API. The helper wraps the `ExportOutput` and adapts input/output for the user and just give users the data in their desired format through `__call__`

However, for internal use as we see here, this was predicted by the original proposal through either 1) Pre-Process transformations or 2) [FXGraphExtractor](https://github.com/pytorch/pytorch/pull/98893). 

We could go with either 1) or 2), but I think 2) is probably the closest component to actually handle user data. `FXGraphExtractor` interface can be extended to add `input_adapter` and `output_adapter` methods that each FX API could implement independently. This way, the `Exporter` never needs to know about any of these IR differences

what do you think?

BowenBao replied:
Not only does different FX extractor may have different set of steps, the fx passes, and onnx passes in the future can too. I view them as "lego pieces" to be used in exporter. 

And as per docstring 

>        The actual formatting steps are associated with each individual export. It
>        depends on the PyTorch model, the particular set of model_args and model_kwargs
>        used for the export, and export options.

the formatting steps are as much "export output" as the model_proto. Hence the design.

thiagocrepaldi replied:
Good point, but I think this might be a different discussion.

The requirement for `ONNX Passes` is that they can either be executed 1) by the exporter after ONNX model is ready to be returned to the user or 2) by the user itself, outside the outside the exporter. In the current proposal, `ExportOutput` will not have the right formatter for scenario 2), as the transformation happened after `ExportOutput` was returned. An external data conversion helper would be needed, right? In fact, that might mean that part of the `ONNX Pass` API may have to "handle" data conversion for the users. Makes sense?

For the `FX Passes`, in general, they are internal and only valid before ONNX export. I believe they should not change the ONNX input/output schema. Otherwise, users are required to always use `ExportOutput`, because the ONNX input/output will not be "predictable" based on original input/output on pytorch eager mode. Makes sense?

BowenBao replied:
That's proving the point with this adapter right? With this design, for signature altering pass, it would be natural to emit adapter as their output along with the model. Again, the adapter serves the purpose to reduce guess work of "what happened to my inputs". If we design passes to operate on `ExportOutput`, then we'd expect the new `ExportOutput` would have internal adapter updated. If we design passes to operate on "Model IR", then original `ExportOutput` is an outdated snapshot.

Unless we restrict `FX Passes` to not change `FX` input/output schema, which is perhaps an interesting topic for discussion but clearly doesn't exist today, it will in turn affects ONNX. 

Bottom line is sure we can ditch this responsibility and have users figure things out as the old exporter did. But since we are pretty much at experimentation stage, I would like to see if we can export the signature adapting info too. In the torch-jit-paritybench benchmark on wild crawled projects and models, the highest percentage failure category for onnx export and validation is signature mismatch, higher than all long tail missing op symbolics combined.

thiagocrepaldi commented on torch/onnx/_internal/fx/dynamo_exporter.py:
```diff
@@ -28,41 +40,206 @@ def export(self) -> torch.onnx.ExportOutput:
+        return self.export_fx_to_onnx(compiler.captured_graph, model_args)
+
+
+class _PyTreeExtensionContext:
```
can you elaborate on why we need to extend pytree?

BowenBao replied:
https://github.com/pytorch/pytorch/issues/98251

thiagocrepaldi replied:
Not for this PR, this almost feels like deserving a deeper discussion. 

If Dynamo cannot support custom types, maybe we could propose an interface for users to register any custom data type, so that exporter knows how to serialize? We can do that at dynamo or exporter context, probably? In fact, we discussed this idea in the past for the torchscript exporter.

BowenBao commented on torch/onnx/_internal/fx/dynamo_exporter.py:
```diff
@@ -28,41 +40,206 @@ def export(self) -> torch.onnx.ExportOutput:
+    # Preserve original function signature.
+    @functools.wraps(model_func)
+    def wrapped(*args, **kwargs):
+        return output_formatter.format(model(*args, **kwargs))
```
good catch

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -155,21 +156,215 @@ def serialize(
+
+@runtime_checkable
+class OutputFormatStep(Protocol):
+    def format(self, model_outputs: Any) -> Any:
```
Because for return type it is basically `Any`, it could be anything. Your proposal is more restrictive. 
```python
def fn(x):
  return x

def fn_return_args_kwargs(x):
  return [x], {}

output_format_step.format(fn(x))  # doesn't work
output_format_step.format(*fn_return_args_kwargs(x))  # works
```

thiagocrepaldi replied:
> Because for return type it is basically `Any`, it could be anything. Your proposal is more restrictive.
> 
> ```python
> def fn(x):
>   return x
> 
> def fn_return_args_kwargs(x):
>   return [x], {}
> 
> output_format_step.format(fn(x))  # doesn't work
> output_format_step.format(*fn_return_args_kwargs(x))  # works
> ```

Given
```python
class InputFormatStep(Protocol):
    def format(
        self, model_args: Sequence[Any], model_kwargs: Mapping[str, Any]
    ) -> Tuple[Sequence[Any], Mapping[str, Any]]:
 ```

wouldn't the example above `output_format_step.format(fn(x))` not work too? `x` is not a `Sequence[Any]`  , right?

BowenBao replied:
Because function arguments can be written as varargs and varkwargs. Return statement cannot.

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -155,21 +156,215 @@ def serialize(
+        ...
+
+
+class InputFormatter:
```
Makes sense

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -155,21 +156,215 @@ def serialize(
+    def __init__(
+        self,
+        model_proto: onnx.ModelProto,
+        input_formatter: InputFormatter,
+        output_formatter: OutputFormatter,
```
> The goal of `ExportOutput` is to hold the exported model.

I think anything specifically associated with a particular export run is reasonable to be counted as "export output". E.g., diagnostic log, formatter. It is pointed out in the other reply that formatter is not a static class/object, but something created by the export call, as is model_proto.

thiagocrepaldi replied:
I am still unsure about it, it really looks a helper function that doesn't need to be created by the exporter.

The `diagnostic` and `model_proto` properties are model+data specific states, unique to that export session execution. One cannot achieve their content through reverse engineering it.

`formatter`, on the other hand, although looks unique to the input data/fx graph extractor used during export, one could easily implement a function without using the exporter to achieve the same thing. It is kind of data conversion recipe that is 1) used by exporter and 2) might be useful for CI/niche users.

For 1), we don't need to expose it publicly. For 2) it is not the most common path, so arguably maybe not too needed by a public API. Especially because there are ways to go around it.

For example,  one has the model `data` (which does not belong to the exporter) and knows which FX graph extractor they want to use, say, `torch.fx.symbolic_trace`, they know how to modify `data` to successfully create the FX graph. `torch.onnx.dynamo_export` and `ExportOutput` are not needed at all. Keeping internalized, decreases public surface, things gets more decoupled and easy to modify later.

BowenBao commented on torch/onnx/_internal/fx/function_dispatcher.py:
```diff
@@ -132,6 +132,7 @@ def aten_getitem(self, i):
     "aten::softmax": ops.special.aten_special_softmax,
     "aten::split": ops.core.aten_split,
     "aten::sqrt": ops.core.aten_sqrt,
+    "aten::stack": ops.core.aten_stack,
```
Shouldn't be here. That's what happens with too many rebases lol

titaiwangms commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -86,95 +168,116 @@ def _run_test_with_fx_to_onnx_exporter_and_onnx_runtime(
+    )
     # This confirms the exported mode accepts different input shapes
     # when dynamic shape is enabled.
+    # TODO(titaiwangms): additional input kwargs.
```
😨

BowenBao replied:
lol, yea forgot to mention this one. We might want to change up the `_run_test_with_fx_to_onnx_exporter_and_onnx_runtime` signature as well. Right now `args` and `kwargs` are taken by a different format.

titaiwangms commented on torch/onnx/_internal/fx/fx_exporter.py:
```diff
@@ -1,23 +1,295 @@
 import torch.onnx._internal.fx.passes as passes
-from torch.onnx._internal import _beartype
+from torch.onnx._internal import _beartype, exporter
+from torch.onnx._internal.fx import fx_exporter
```
Is this file importing itself?

BowenBao replied:
great catch, how is this not failing ..

thiagocrepaldi replied:
`mypy` is letting me down too, even after `make setup_lint` and `lintrunner init` and finally `lintrunner`. CI finds a bunch of failures while local runs pass

titaiwangms commented on torch/onnx/_internal/fx/fx_exporter.py:
```diff
@@ -63,6 +326,17 @@ def export_fx_to_onnx(
+            )
+            # ONNX can't represent collection types (e.g., dictionary, tuple of tuple of
+            # tensor, etc), we flatten the collection and register each element as output.
+            self._output_formatter.append_step(fx_exporter.FlattenOutputStep())
```
I wonder if you imagine steps working kind of like passes, should we create another folder for steps.

titaiwangms commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -360,6 +571,15 @@ def forward(self, x, y):
 
+    @pytorch_test_common.skip_fx_exporters(
+        {
+            DynamoExporter: (
```
I didn't even realize we weren't using `DynamoExporter`...

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -190,14 +385,8 @@ def save(
-            result, _ = _pytree.tree_flatten(self.actual_forward(*args, **kwargs))
-            return result
+    _input_formatter: InputFormatter
+    _output_formatter: OutputFormatter
```
It's final for `ExportOutput`. For `Exporter`, these are built up during export. But I think you helped me found an edge case of repeated exports..

justinchuby commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -244,6 +367,94 @@ def func(x, b=torch.tensor(1.0)):
+    # beartype.roar.BeartypeCallHintParamViolation:
+    # @beartyped onnxscript.function_libs.torch_aten.graph_building.TorchScriptGraph.add_input()
+    # parameter input_value=8.0 violates type hint typing.Union[torch.Tensor, NoneType],
+    # as float 8.0 not <class "builtins.NoneType"> or <protocol "torch.Tensor">.
```
Hmm. Should we wrap it as a tensor?

BowenBao replied:
I thought in general this is the problem of having number type, instead of tensor type, as input for onnx. There is not fundamentally blocking, just not covered yet.

BTW how is 5.0 different than 8.0? Did you mean `tensor(5.0)`?

titaiwangms replied:
Oh NVM. I was confused by the previous test case.

titaiwangms replied:
> Hmm. Should we wrap it as a tensor?

I think so. We expects shape/type from inputs.

thiagocrepaldi replied:
Could we do something like `is_builtin_type(input_value)` and proceed with either its tensor.dtype or scalar type (built-in type)

Although unusual, builtin type is present in models; `bool` should be the most common, but users are free to input other than `Union[torch.Tensor, NoneType]`

thiagocrepaldi replied:
> Hmm. Should we wrap it as a tensor?

I think this may have side effects. That is why Meta created `SymInt` and `SymFloat`. wrapping as tensor may change the execution path

titaiwangms replied:
I think the error message seems outdated, as the add_input API has been changed for a while. For SymInt and SymFloat, I think it should be fine if we wrap them as tensor, as we are already doing it in dynamic exporting to enable their calculations(add, multiple, divide, ...). Looking forard to playing around with this test case.

BowenBao replied:
Updated error message. Follow up on builtin types separately.

BowenBao committed:
Address feedbacks with small fixes on "[ONNX] Introduce Input/Ouptut formatter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output formatter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output formatter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

BowenBao committed:
Rename from format to adapt on "[ONNX] Introduce Input/Ouptut formatter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output formatter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output formatter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

titaiwangms commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -207,25 +316,39 @@ def func(x):
+        {
+            DynamoExporter: (
+                "AssertionError: Dynamo input/output is not consistent with traced input/output. "
+                "Ref: https://github.com/pytorch/pytorch/issues/96379"
```
I think the status of `test_arange` and `test_scalar_tensor` should be updated with this PR.

BowenBao replied:
Updated.

BowenBao committed:
Update test skips; add a few missed 'adapt' renaming on "[ONNX] Introduce Input/Ouptut formatter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output formatter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output formatter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -155,21 +156,215 @@ def serialize(
 
+# TODO(bowbao): Add diagnostics for IO formatters.
+@runtime_checkable
+class InputFormatStep(Protocol):
```
added docstring

thiagocrepaldi commented on torch/onnx/_internal/fx/fx_exporter.py:
```diff
@@ -1,23 +1,297 @@
+        self._input_adapter.append_step(adapt_step)
+        return adapt_step.adapt(model_args, model_kwargs)
+
+    def _apply_output_adapt_step(
```
we could move all these adapters to a file of its own and declutter the exporter class files.

BowenBao replied:
Sure for follow ups

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -155,21 +156,241 @@ def serialize(
+        self._input_adapt_steps.append(step)
+
+    @_beartype.beartype
+    def to_onnx(self, *model_args, **model_kwargs) -> Sequence[torch.Tensor]:
```
```suggestion
    def apply(self, *model_args, **model_kwargs) -> Sequence[torch.Tensor]:
```
Evrry time that I read this, I expect to see a `ModelProto` lol. There is nothing in this class that restricts it to onnx, so we could use a generic term such as `apply`

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -155,21 +156,241 @@ def serialize(
+    input signature by any individual component in the exporter.
+    """
+
+    def adapt(
```
```suggestion
    def apply(
```

BowenBao commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -244,6 +367,109 @@ def func(x, b=torch.tensor(1.0)):
+        )
+
+    @pytorch_test_common.skip_fx_exporters(
+        {DynamoOptimizeExporter: "Unhandled used argument."}
```
Because this PR is formalizing what exists today. It does not go in and fix actual problems.

BowenBao replied:
thanks for helping spotting the typo -> unused lol

BowenBao commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -244,6 +367,109 @@ def func(x, b=torch.tensor(1.0)):
+            DynamoExporter: (
+                "beartype.roar.BeartypeCallHintReturnViolation: @beartyped "
+                "torch.onnx._internal.exporter.ExportOutput.adapt_torch_inputs_to_onnx() "
+                "return (tensor([[[ 1.5410, -0.2934]]]), 8.0) violates type hint "
```
We can, it can happen in follow ups though. I'm unsure yet if we'd need adjustments on fx_to_onnx conversion pass, so don't want to extend the scope of this PR.

BowenBao commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -244,6 +367,109 @@ def func(x, b=torch.tensor(1.0)):
+                "instance of <protocol 'torch.Tensor'>."
+            ),
+            DynamoOptimizeExporter: (
+                "RuntimeError: The two modules have different number of arguments. "
```
Same as the other reply.

BowenBao committed:
rebase + final feedbacks on "[ONNX] Introduce Input/Ouptut adapter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output adapter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output adapter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Introduce Input/Ouptut adapter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output adapter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output adapter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Introduce Input/Ouptut adapter; Switch to 'DynamoExporter'"


Summary
* Introduce input/output adapter. Due to design differences, input/output format
between PyTorch model and exported ONNX model are often not the same. E.g., `None`
inputs are allowed for PyTorch model, but are not supported by ONNX. Nested constructs
of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX,
etc. The new input/output adapter is exported with the model. Providing an interface to
automatically convert and validate inputs/outputs format.
* As suggested by #98251, 
provide extension for unwrapping user defined python classes for `dynamo.export` based
exporter. Unblock huggingface models.
* Re-wire tests to run through `DynamoExporter` w/ `dynamo_export` api. Kept 
`DynamoOptimizeExporter` in the tests for now for coverage of this change.

[ghstack-poisoned]
